File: generator/validators.py
from generators import valueOrRowValue


def validateNumber(value, rule):
    operator = rule["operator"]
    ruleValue = float(rule["value"])

    if operator == "eq":
        return value == ruleValue

    if operator == "gt":
        return value > ruleValue

    if operator == "gte":
        return value >= ruleValue

    if operator == "lt":
        return value < ruleValue

    if operator == "lte":
        return value <= ruleValue

    return False


def ruleApplies(row, schema, rule):
    fieldName = rule["field"]
    fieldValue = row[fieldName]
    fieldSchema = schema[fieldName]
    valueType = fieldSchema["type"]

    if valueType == "number":
        return validateNumber(float(fieldValue), rule)

    return fieldValue == rule["value"]


def validateValue(row, schema, fieldName):
    value = row[fieldName]
    fieldSchema = schema[fieldName]
    rules = fieldSchema.get("rules")

    if not rules:
        return value

    for rule in rules:
        if ruleApplies(row, schema, rule):
            return valueOrRowValue(rule["result"], row)

    return value

File: app/services/validators.js
const numberRuleApplies = ({ fieldValue, rule }) => {
  const ruleOperator = rule.operator ?? "eq";
  const ruleValue = rule.value;

  if (ruleOperator === "eq") {
    return fieldValue === ruleValue;
  }

  if (ruleOperator === "gt") {
    return fieldValue > ruleValue;
  }

  if (ruleOperator === "gte") {
    return fieldValue >= ruleValue;
  }

  if (ruleOperator === "lt") {
    return fieldValue < ruleValue;
  }

  if (ruleOperator === "lte") {
    return fieldValue <= ruleValue;
  }

  return false;
};

const ruleApplies = ({ row, schema, rule }) => {
  const fieldValue = row[rule.field];
  const fieldSchema = schema[rule.field];

  if (fieldSchema.value === "number") {
    return numberRuleApplies({ fieldValue, rule });
  }

  return fieldValue === rule.value;
};

export const validateValue = ({ row, schema, fieldName }) => {
  const rowValue = row[fieldName];
  const fieldSchema = schema[fieldName];
  const rules = fieldSchema.rules ?? [];

  if (!rules.length) {
    return rowValue;
  }

  const applicableRule = rules.find((rule) =>
    ruleApplies({ row, schema, rule })
  );

  return applicableRule?.result ?? rowValue;
};
